// indicative-rules/build/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const date_fns_1 = require("date-fns");
const months = {
    years: (unit) => unit * 12,
    quarters: (unit) => unit * 3,
    months: (unit) => unit,
};
const days = {
    weeks: (unit) => unit * 7,
    days: (unit) => unit,
};
const milliseconds = {
    hours: (unit) => unit * 3600000,
    minutes: (unit) => unit * 60000,
    seconds: (unit) => unit * 1000,
    milliseconds: (unit) => unit,
};
exports.calcUnits = (diffUnit, key, operator) => {
    if (months[key]) {
        return date_fns_1.addMonths(new Date(), operator === '-' ? -months[key](diffUnit) : months[key](diffUnit));
    }
    if (days[key]) {
        return date_fns_1.addDays(new Date(), operator === '-' ? -days[key](diffUnit) : days[key](diffUnit));
    }
    if (milliseconds[key]) {
        return date_fns_1.addMilliseconds(new Date(), operator === '-' ? -milliseconds[key](diffUnit) : milliseconds[key](diffUnit));
    }
};
exports.dateOffsetKeys = [
    'years',
    'quarters',
    'months',
    'weeks',
    'days',
    'hours',
    'minutes',
    'seconds',
    'milliseconds',
];
